src/domain/runner/store.rs

```diff
@@ -0,0 +1,387 @@
+        self.values.contains_key(pointer)
+    }
+
+    pub fn get<GetFun>(
+        &mut self,
+        pointer: &[u8; STORAGE_POINTER_SIZE],
+        get_fn: GetFun,
+    ) -> Result<CacheResponse, RuntimeError>
+    where
+        GetFun: Fn(StoragePointer) -> Result<StorageValue, RuntimeError>,
+    {
+        self.reads += 1;
+
+        // If object exists in the cache
+        if let Some(value) = self.values.get(pointer) {
+            Ok(CacheResponse::new(value.current, LOAD_WARM, 0))
+        }
+        // first acces of pointer
+        else {
+            let original_value = get_fn(*pointer)?;
+            self.values
+                .insert(*pointer, CacheValue::new(original_value));
+
+            Ok(CacheResponse::new(original_value, LOAD_COLD, 0))
+        }
+    }
+
+    pub fn set<GetFun, SetFun>(
+        &mut self,
+        key: [u8; STORAGE_POINTER_SIZE],
+        value: [u8; STORAGE_VALUE_SIZE],
+        get_value: GetFun,
+        set_value: SetFun,
+    ) -> Result<CacheResponse, RuntimeError>
+    where
+        GetFun: Fn(StoragePointer) -> Result<StorageValue, RuntimeError>,
+        SetFun: Fn(StoragePointer, StorageValue) -> Result<(), RuntimeError>,
+    {
+        // increase number of writes
+        self.writes += 1;
+
+        let mut gas_cost = 0;
+        let mut gas_refund: i64 = 0;
+
+        let mut cache_value = if let Some(value) = self.values.get(&key) {
+            // Warm access
+            value.clone()
+        } else {
+            // Cold access
+            gas_cost += LOAD_COLD;
+            CacheValue::new(get_value(key)?)
+        };
+
+        if value == cache_value.current {
+            // No changes
+            gas_cost += STORE_BASE;
+        } else {
+            // Clean slot (not yet updated in current execution context)
+            if cache_value.original == cache_value.current {
+                // (slot started zero, currently still zero, now being changed to nonzero)
+                if cache_value.original == STORAGE_VALUE_ZERO {
+                    gas_cost += STORE_NEW;
+                }
+                // (slot started nonzero, currently still same nonzero value, now being changed):
+                else {
+                    gas_cost += STORE_UPDATE;
+                }
+            } else {
+                gas_cost += STORE_BASE
+            };
+
+            if cache_value.current == cache_value.original {
+                if cache_value.original != STORAGE_VALUE_ZERO && value == STORAGE_VALUE_ZERO {
+                    gas_refund = STORE_REFUND as i64;
+                }
+            } else {
+                if cache_value.original != STORAGE_VALUE_ZERO {
+                    if cache_value.current == STORAGE_VALUE_ZERO {
+                        // TODO: Add to gas???
+                        if value != cache_value.original {
+                            gas_refund -= STORE_REFUND as i64;
+                        } else {
```

Comment:
This section is different from the formula on:
https://www.evm.codes/?fork=cancun#55

But values are the same as calculation (times 1_000)